File: python/pyxis/_datum_compensate.py
from typing import overload

import numpy as np
from numpy.typing import NDArray

from ._utils import TCoordArray, TCoordScalar, coord_util


@overload
def datum_compensate(
    xc: TCoordScalar,
    yc: TCoordScalar,
    hb: float,
    radius: float,
    x0: float,
    y0: float,
    clone: bool = True,
) -> tuple[float, float]: ...
@overload
def datum_compensate(
    xc: TCoordArray,
    yc: TCoordArray,
    hb: float,
    radius: float,
    x0: float,
    y0: float,
    clone: bool = True,
) -> tuple[NDArray[np.float64], NDArray[np.float64]]: ...
def datum_compensate(
    xc,
    yc,
    hb: float,
    radius: float,
    x0: float,
    y0: float,
    clone: bool = True,
):
    r"""
    Convert projected XY coordinates from height compensation plane to sea level plane.

    Unit: meter

    Parameters
    ----------
    xc, yc
        Coordinates on height compensation plane.
    hb
        Elevation of height compensation plane.
    radius
        Radius of earth.
    x0, y0
        Coordinate system origin.
    clone
        If True, a deep copy of `data` will be created before processing.
        If False, `data` will be modified in place. Default is False.

    Returns
    -------
    tuple[Any, Any]
        Projected XY coordinates of sea level plane.

    References
    ----------
    - 杨元兴.抵偿高程面的选择与计算[J].城市勘测,2008(02):72-74.
    """
    from .pyxis_py import py_datum_compensate  # type: ignore

    xc = coord_util("xc", xc, clone)
    yc = coord_util("yc", yc, clone)

    xc, yc = py_datum_compensate(xc, yc, hb, radius, x0, y0)

    return xc, yc

File: rust/crates/pyxis/src/datum_compensate.rs
use crate::GeoFloat;
/// Converts projected XY coordinates from the height compensation plane to
/// the sea level plane.
///
/// # Arguments
///
/// - `xc`, `yc`: Coordinates on the height compensation plane (in meters).
/// - `hb`: Elevation of the height compensation plane (in meters).
/// - `radius`: Radius of the Earth (in meters).
/// - `x0`, `y0`: Coordinate system origin (in meters).
///
/// # Returns
///
/// A tuple containing the projected XY coordinates of the sea level plane
/// (in meters).
///
///
/// # References
/// - 杨元兴. (2008). 抵偿高程面的选择与计算. 城市勘测 (02), 72-74.
///
/// # Examples
/// ```
/// use float_cmp::assert_approx_eq;
/// let p = (469704.6693, 2821940.796);
/// let processor = pyxis::DatumCompensate::new(400.0, 6_378_137.0, 500_000.0, 0.0);
/// let p = processor.transfrom(p.0, p.1);
/// assert_approx_eq!(f64, p.0, 469706.56912942487, epsilon = 1e-17);
/// assert_approx_eq!(f64, p.1, 2821763.831232311, epsilon = 1e-17);
/// ```
pub struct DatumCompensate<T: GeoFloat> {
    x0: T,
    y0: T,
    factor: T,
}
impl<T: GeoFloat> DatumCompensate<T> {
    pub fn new(hb: T, radius: T, x0: T, y0: T) -> Self {
        let ratio = hb / radius;
        let factor = ratio / (T::ONE + ratio);
        Self { x0, y0, factor }
    }

    pub fn transfrom(&self, xc: T, yc: T) -> (T, T)
    where
        T: GeoFloat,
    {
        let xc = xc - self.factor * (xc - self.x0);
        let yc = yc - self.factor * (yc - self.y0);
        (xc, yc)
    }
}
